upup/pkg/fi/cloudup/scaleway/cloud.go

```diff
@@ -16,7 +16,393 @@ limitations under the License.
+	DeregisterInstance(instance *cloudinstances.CloudInstance) error
+	DeleteInstance(i *cloudinstances.CloudInstance) error
+
+	GetClusterServers(clusterName string, serverName *string) ([]*instance.Server, error)
+	GetClusterVolumes(clusterName string) ([]*instance.Volume, error)
+
+	DeleteServer(server *instance.Server) error
+	DeleteVolume(volume *instance.Volume) error
+}
+
+// static compile time check to validate ScwCloud's fi.Cloud Interface.
+var _ fi.Cloud = &scwCloudImplementation{}
+
+// scwCloudImplementation holds the scw.Client object to interact with Scaleway resources.
+type scwCloudImplementation struct {
+	client *scw.Client
+	region scw.Region
+	zone   scw.Zone
+	tags   map[string]string
+
+	accountAPI  *account.API
+	instanceAPI *instance.API
+}
+
+// NewScwCloud returns a Cloud with a Scaleway Client using the env vars SCW_ACCESS_KEY, SCW_SECRET_KEY,
+// SCW_DEFAULT_PROJECT_ID, SCW_DEFAULT_REGION and SCW_DEFAULT_ZONE
+func NewScwCloud(tags map[string]string) (ScwCloud, error) {
+	region, err := scw.ParseRegion(os.Getenv("SCW_DEFAULT_REGION"))
+	if err != nil {
+		return nil, fmt.Errorf("error parsing SCW_DEFAULT_REGION: %w", err)
+	}
+	zone, err := scw.ParseZone(os.Getenv("SCW_DEFAULT_ZONE"))
+	if err != nil {
+		return nil, fmt.Errorf("error parsing SCW_DEFAULT_ZONE: %w", err)
+	}
+
+	// We make sure that the credentials env vars are defined
+	scwAccessKey := os.Getenv("SCW_ACCESS_KEY")
+	if scwAccessKey == "" {
+		return nil, fmt.Errorf("SCW_ACCESS_KEY has to be set as an environment variable")
+	}
+	scwSecretKey := os.Getenv("SCW_SECRET_KEY")
+	if scwSecretKey == "" {
+		return nil, fmt.Errorf("SCW_SECRET_KEY has to be set as an environment variable")
+	}
+	scwProjectID := os.Getenv("SCW_DEFAULT_PROJECT_ID")
+	if scwProjectID == "" {
+		return nil, fmt.Errorf("SCW_DEFAULT_PROJECT_ID has to be set as an environment variable")
+	}
+
+	scwClient, err := scw.NewClient(
+		scw.WithUserAgent("kubernetes-kops/"+kopsv.Version),
+		scw.WithEnv(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("error building client for Scaleway Cloud: %w", err)
+	}
+
+	return &scwCloudImplementation{
+		client:      scwClient,
+		region:      region,
+		zone:        zone,
+		tags:        tags,
+		accountAPI:  account.NewAPI(scwClient),
+		instanceAPI: instance.NewAPI(scwClient),
+	}, nil
+}
+
+func (s *scwCloudImplementation) Region() string {
+	return string(s.region)
+}
+
+func (s *scwCloudImplementation) Zone() string {
+	return string(s.zone)
+}
+
+func (s *scwCloudImplementation) ClusterName(tags []string) string {
+	for _, tag := range tags {
+		if strings.HasPrefix(tag, TagClusterName) {
+			return strings.TrimPrefix(tag, TagClusterName+"=")
+		}
+	}
+	return ""
+}
+
+func (s *scwCloudImplementation) ProviderID() kops.CloudProviderID {
+	return kops.CloudProviderScaleway
+}
+
+func (s *scwCloudImplementation) DNS() (dnsprovider.Interface, error) {
+	//TODO(Mia-Cross) implement me
+	panic("Scaleway doesn't have a DNS yet")
```

Comment:
Remove the TODO. What if we don't implement this function? it does not seem to be required by `fi.CloudInterface`

Reply:
The DNS function is indeed required by the fi.Cloud interface. I removed the TODO but it will not stay unimplemented for long since the DNS part is already finished on an other branch.